RFC 1: Moving Funds Between Wallets (#127)
keep-network/tbtc-v2: docs/rfc/rfc-1.adoc

of being able to meet the 51-of-100 threshold to produce signatures.
Additionally, we want to avoid situations where operators are the custodians of
a wallet for extended periods. To avoid these issues, once a wallet is older
than the `wallet_max_age`, or if it drops below the liveness threshold (say,
below 70 on a <<heartbeat,heartbeat>>), we motion to
<<closing-a-wallet,transfer the funds>> to another randomly selected wallet.

Once a wallet no longer has funds and is not the primary wallet for new
deposits, it can be <<closing-a-wallet,closed>> and operators are no longer
required to maintain it.

[[closing-a-wallet]]

> ==== Closing A Wallet

Reviewer: FWIW, we already have the "Report heartbeat failed" transaction listed in RFC 6. The only one we are missing is the transaction submitting the proof of moving the funds. I think it should land in the Maintainer transaction incentives section.

Governable Parameters:

- `wallet_min_btc`: The smallest amount of BTC a wallet can hold before we
  attempt to close the wallet and transfer the funds to a randomly selected

>   wallet.

Author: > Maybe we will only perform that check after redemptions? That would make a bit more sense, but still somebody could make a deposit on a fresh wallet and make a redemption soon afterwards.

Yeah - that's the intention! The definition is weird out of context in a list, so I'll add some clarity.

As for the second point, active wallets can't be closed, and redemptions are still supposed to be coming from the oldest wallets (generally).

-----

Say that `wallet_min_btc = 1 BTC`. If a 3-week old wallet (no longer active) has 1.5 BTC left, and someone skips the dApp to redeem against that wallet instead of the oldest (and we allow that in our validation contract), then it would fall below the threshold and should close. I think this is okay! Closing that wallet is better than making 100 people look after a 0.5 BTC wallet until all of the older wallets get around to finally clearing out.

>   wallet.

Author: clarified in [0ea9675](https://github.com/keep-network/tbtc-v2/pull/127/commits/0ea96758d96f51a03c7df587d3485964b61e2e5f)

> When a wallet fails a <<heartbeat,heartbeat>> or ends a

Author: addressed in [bae7e6c](https://github.com/keep-network/tbtc-v2/pull/127/commits/bae7e6c61b6028eb155321728433cd5b7a320e59)

> <<redemption,redemption>> with less than `wallet_min_btc` remaining, the

Author: addressed in [0704bae](https://github.com/keep-network/tbtc-v2/pull/127/commits/0704baebb3cdf5e953a6e6c4f41f4de939a80a3c)

operators begin the process of closing it. If there are still funds remaining,
the <<operator-only,first operator>> posts a <<operator-only,reimbursable>>
transaction to the ethereum chain declaring an intention to close the wallet.

That operator proposes a BTC fee (as in <<sweeping,sweeping>>) and an ethereum

> block height commitment, and they use that to construct a P2PKH transaction

Author: rewrote the whole section in [e4717ca](https://github.com/keep-network/tbtc-v2/pull/127/commits/e4717caa5fa2862dce4f2e184f670b441c6fc43b)

moving the wallet's main UTXO to another randomishly chosen wallet where the
source of randomness comes from the ethereum block hash at the pre-commited

> block height.

Reviewer: I do not understand the pre-committed block height, can you please elaborate?

What I think we could do instead of an additional commitment is that when someone (does not even have to be an operator from that wallet - this is *public-knowledge*) notifies that a wallet has funds below the dust threshold, they submit a transaction to the chain to select another wallet where the funds should go to.

We take the blockhash (I think this is sufficient) and % N where N is the number of most active wallets we are interested in. We record blockhash % N as the target wallet assuming the amount hold by that wallet is above the dust threshold or that wallet is active; if not we decrement to move to the more recent wallets.

For example:

wallet 20 [ active ] 0.01 BTC
wallet 19 [ active ] 200 BTC
wallet 18 [ inactive ] 198 BTC
wallet 17 [inactive ] 194 BTC
...
wallet 3 [ inactive ] 10 BTC
wallet 2 [ inactive ] 5 BTC
wallet 1 [ inactive ] 0.01 BTC

With N = 4, we would be selecting from wallets 20-17 and all wallets qualify; they are either above the dust threshold (17-19) or they are below the dust threshold but active (20).

Author: I like this scheme - so the public just *points out* after a redemption that the funds need to be transferred, or the operators *point out* after a failed heartbeat that the funds need to be transferred. No one but the miners are able to change the hash of this "hey we need to transfer" declaration, and it carries two benefits:

a) it determines the target wallet on-chain
b) it sets the beginning of the time-out on-chain

:+1:

Author: addressed in [e4717ca](https://github.com/keep-network/tbtc-v2/pull/127/commits/e4717caa5fa2862dce4f2e184f670b441c6fc43b)

After this transaction is complete, <<public-knowledge,the public>> is able
to submit a SPV proof to let ethereum know that the funds were transferred
nonfraudulently, and which account balances need to be updated (like the

> treasur which pays the bitcoin mining fee).

Author: addressed in [8699428](https://github.com/keep-network/tbtc-v2/pull/127/commits/869942844909fd992c7370857c45db7cd9459132)

*Notes*: We trust the honest majority to pick a new random wallet. Transferring
the BTC to any address other than the P2PKH of one of the other wallet
addresses is fraud, and is <<Punishment,punishable>>. Moving funds without

> sending a follow-up proof is <<Punishment,punishable>>. Failure to close a

Reviewer 2: So the follow-up proof can be sent by anyone and if it's not sent (in what time?), the operators of the emptying wallet may be punished. So usually it will be operators of the wallet who send that proof (it's in their interest to do that), but theoretically this could be done by everybody, because everybody has a sufficient knowledge to do that. Do I understand correctly?

Author: e4717caa5fa2862dce4f2e184f670b441c6fc43b fleshes out the timeout mechanism here. We post a transaction when we notice that a wallet needs to be closed (anyone including operators can do it), and we use the hash of that tx to decide which wallet to send funds to. The timeout starts counting from when that tx clears.

Other than that, the rest of your understanding is spot on!

> wallet that failed a <<heartbeat,heartbeat>> or fell below `wallet_min_btc` is

Reviewer 2: ```suggestion
wallet that failed a <<heartbeat,heartbeat>>, fell below `wallet_min_btc`, or exceeded `wallet_max_age` is
```

Author: addressed in [0704bae](https://github.com/keep-network/tbtc-v2/pull/127/commits/0704baebb3cdf5e953a6e6c4f41f4de939a80a3c)

> <<Punishment,punishable>>.

Reviewer: We can make it a bit more strict. When someone proposes to move the funds to another wallet, we record the target wallet on-chain. Then, non-fraudulent redemption is the one following rules from https://github.com/keep-network/tbtc-v2/issues/119 or the one that moves funds to the wallet recorded in the transaction requesting the move of funds.

Author: addressed in [e4717ca](https://github.com/keep-network/tbtc-v2/pull/127/commits/e4717caa5fa2862dce4f2e184f670b441c6fc43b)

> <<Punishment,punishable>>.

Reviewer 2: All those punishment transactions would lead to slashing, do I get that right?

Author: Yeah - everything listed here is slashable

deposits, it can be <<closing-a-wallet,closed>> and operators are no longer
required to maintain it.

[[closing-a-wallet]]
==== Closing A Wallet
Governable Parameters:

- `heartbeat`: The number of group members required for a
  <<heartbeat,heartbeat>> to be successful.
- `wallet_closure_timeout`: The amount of time that a wallet has to
  successfully move its funds to another wallet and inform the ethereum chain
  before it is at risk of punishment.
- `wallet_max_age`: The oldest we allow a wallet to become before we transfer the funds
  to a randomly selected wallet.
- `wallet_min_btc`: The smallest amount of btc a wallet can hold after a
  redemption before we attempt to close the wallet and transfer the funds to a
  randomly selected wallet.

When a wallet drops below the `heartbeat` threshold during a
<<heartbeat,heartbeat>> or ends a <<redemption,redemption>> with less than
`wallet_min_btc` remaining, or exceeds the `wallet_max_age`, the operators
begin the process of closing it. If there are still funds remaining, the
<<operator-only,first operator>> posts a <<operator-only,reimbursable>>
transaction to the ethereum chain declaring an intention to close the wallet.

> That operator proposes a BTC fee (as in <<sweeping,sweeping>>) and, and the

Author: [aa5fad1](https://github.com/keep-network/tbtc-v2/pull/127/commits/aa5fad1759eccd49630610dddd1625faf483d1f1)

operators use the hash of the previous intention-to-close transaction as a
source of randomness to choose viable (it must also have a passing heartbeat

> and be above `wallet_min_btc`) to transfer the remaining funds to. They

Author: [aa5fad1](https://github.com/keep-network/tbtc-v2/pull/127/commits/aa5fad1759eccd49630610dddd1625faf483d1f1)

  successfully move its funds to another wallet and inform the ethereum chain
  before it is at risk of punishment.
- `wallet_max_age`: The oldest we allow a wallet to become before we transfer the funds
  to a randomly selected wallet.
- `wallet_min_btc`: The smallest amount of btc a wallet can hold after a
  redemption before we attempt to close the wallet and transfer the funds to a
  randomly selected wallet.

When a wallet drops below the `heartbeat` threshold during a
<<heartbeat,heartbeat>> or ends a <<redemption,redemption>> with less than
`wallet_min_btc` remaining, or exceeds the `wallet_max_age`, the operators
begin the process of closing it. If there are still funds remaining, the
<<operator-only,first operator>> posts a <<operator-only,reimbursable>>
transaction to the ethereum chain declaring an intention to close the wallet.

That operator proposes a BTC fee (as in <<sweeping,sweeping>>) and the
operators use the hash of the previous intention-to-close transaction as a
source of randomness to choose viable (it must also have a passing heartbeat
and be above `wallet_min_btc`) wallet to transfer the remaining funds to. They

> construct a P2PKH transaction moving the wallet's main UTXO to that wallet.

Author: okay - [728e0e5](https://github.com/keep-network/tbtc-v2/pull/127/commits/728e0e5d1f1c7443167b8cde03c756eeb78ddec8)

  before it is at risk of punishment.
- `wallet_max_age`: The oldest we allow a wallet to become before we transfer the funds
  to a randomly selected wallet.
- `wallet_min_btc`: The smallest amount of btc a wallet can hold after a
  redemption before we attempt to close the wallet and transfer the funds to a
  randomly selected wallet.

When a wallet drops below the `heartbeat` threshold during a
<<heartbeat,heartbeat>> or ends a <<redemption,redemption>> with less than
`wallet_min_btc` remaining, or exceeds the `wallet_max_age`, the operators
begin the process of closing it. If there are still funds remaining, the
<<operator-only,first operator>> posts a <<operator-only,reimbursable>>
transaction to the ethereum chain declaring an intention to close the wallet.

That operator proposes a BTC fee (as in <<sweeping,sweeping>>) and the
operators use the hash of the previous intention-to-close transaction as a
source of randomness to choose viable (it must also have a passing heartbeat
and be above `wallet_min_btc`, and be within the `random_transfer_depth`)
wallet to transfer the remaining funds to. They construct a P2PKH transaction

> moving the wallet's main UTXO to that wallet.

Reviewer: We [can not](https://docs.soliditylang.org/en/latest/units-and-global-variables.html) have the hash of the previous intention-to-close transaction on-chain so we will not be able to validate on-chain if the funds were moved to the right wallet. If we do not have to validate it on-chain and we can trust the majority in the wallet to move the funds to any of the `random_transfer_depth` new wallets, this is fine. But if we want to validate it on-chain, then we need to choose something else as the seed.

Options that come to my mind:
- ask the random beacon
- use the most recently created wallet's pubkey hash
- use relay entry from the latest DKG

All these options have their pros and cons. Random beacon is probably the most secure one but it adds some complexity given that the beacon can be busy at the moment of failing the heartbeat and we do want to start moving funds immediately.

On the first sight, hash of the pubkey of the most recently created wallet does not sound that bad but worth thinking more about it cc @eth-r

Reviewer: The transaction hash is not available from Solidity function, see the parameters [here](https://docs.soliditylang.org/en/latest/units-and-global-variables.html). We can have it off-chain but not on-chain.

Author: I think it's fine that we just have it off-chain, since we're just using the information as a way to short-cut consensus.

Here's how I imagine this plays out in the happy path:

+ We detect that we need to transfer funds, and that gets posted to ethereum with a particular hash
+ The wallet reads that message, uses that hash off-chain to all seed the same RNG and pick a uniform random number between [0, `random_transfer_depth`], and picks a wallet from that number, and transfers the funds to that wallet
+ No one challenges the tx because it isn't fraudulent

Unchallengable Bad Path

+ We detect that we need to transfer funds, and that gets posted to ethereum with a particular hash
+ The wallet reads that message, colludes to send it to a wallet within [0, `random_transfer_depth`], and transfers the funds to that wallet
+ No one challenges the tx because it doesn't look fraudulent

(this requires a 51-of-100 attack with modified clients)

Challengable Bad Path

+ We detect that we need to transfer funds, and that gets posted to ethereum with a particular hash
+ The wallet reads that message, colludes to send it to a wallet outside of [0, `random_transfer_depth`], and transfers the funds to that wallet (could be a totally separate wallet)
+ challenge/response sighash etc fraud-proof process

(this requires a 51-of-100 attack with modified clients)

---

So, we open ourselves up to a 51-of-100 attack where the wallet is able to move the funds to a wallet *that they already control* to another wallet of their choosing within `[0, random_transfer_depth]` and have it not seem like fraud.

---

Asking the random beacon sounds gas expensive, using the wallet pubkey hash is past, static information which means that all wallet closures would go to the same wallet until a new wallet is created unless you also had some sort of indexing, and then folks could manipulate the index to get the funds to move into the wallet they wanted.

Can you elaborate on what the relay entry is?

So my main question at the moment is: how expensive is the attack I'm describing above vs paying the random beacon to prevent that attack? There's a decent chance that we're overengineering this at the moment and we could just do the off-chain consensus based on the intention-to-close transaction and upgrade to random-beacon later after more analysis later

After this transaction is complete, <<public-knowledge,the public>> is able to
submit a SPV proof to let ethereum know that the funds were transferred
nonfraudulently, and which account balances need to be updated (like the

> treasury which pays the bitcoin mining fee).

Reviewer: I think we should do the bookkeeping of Bitcoin fees off-chain to minimise gas costs. I think it is enough to emit an event with Bitcoin transaction hash so that the treasury governance can do the donation (see https://github.com/keep-network/tbtc-v2/pull/132/files#diff-1044c578112615d9cb745f17e8203a444e72ddb3c5d8d5fd97d97038952c77b1R622) after some time.

Author: addressed in 73c7f373212d240c743d6664dbcc01342116dbd9

*Notes*: Transferring the BTC to any address other than the P2PKH of one of the
other wallet addresses is fraud, and is <<Punishment,punishable>>. Moving funds
without sending a follow-up proof is <<Punishment,punishable>>. Failure to
close a wallet that failed a <<heartbeat,heartbeat>> or fell below

> `wallet_min_btc`, or exceeded `wallet_max_age` is <<Punishment,punishable>>.

Reviewer: But how do we know that on-chain? It feels like it's just a lack of notification about the current state from maintainers.

Author: What I mean here is that we want to be in a state where we have maintainers that are incentivized to post transactions about the state of wallets (they're too old, too low on funds, they failed a heartbeat, etc). Once one of those transactions gets posted, it starts a timeout. If the close-the-wallet process isn't complete before the time runs out, the wallet gets punished.

Does that answer your question, or are you asking about how we get the maintainers to keep the state up to date?

[[transaction-incentives]]
=== Transaction Incentives
Transaction incentives are more deeply explored in link:rfc-6.adoc[RFC 6:
Transaction Incentives]. Summarized:

There are three different types of transactions: <<operator-only,Operator-Only>>,
<<public-knowledge,Public-Knowledge>>, and <<punishment,Punishment>>.

[[operator-only]]
==== Operator-Only
Operator-Only transactions are where only the operators have access to the
information required to assemble the transaction with the right input
parameters.

In order to avoid all operators racing to submit the

> transaction at the same time, we have an off-chain gentleman's agreement to

Author: I think the final consensus here was a meme 

![image](https://user-images.githubusercontent.com/1045160/152419496-d47bacd7-2553-4fb4-97df-f4623eeaffb3.png)

Jokes aside, briefly checking the internet grammar wizards out there, I think "informal" is probably the most accurate and inclusive. I'm down to standardize on that :D

> transaction at the same time, we have an off-chain gentleman's agreement to

Author: I also think that standardization should happen in #125 so that this PR doesn't get muddied up

submit based on the operator's position in the group (can use the hash of the
group's pubkey).

If the designated operator does not submit their transaction before a timeout
expires, the duty moves to the next operator and the group can sign a
transaction to mark that operator as inactive. Such a transaction would be a

> <<punishment,Punishment>>.

Reviewer 2: I just want to make sure I understand correctly (plus have one question):
Let's say we're talking about transaction submitting DKG result. So in the case where the first operator does not submit the DKG result before the timeout, second operator would send that transaction. This transaction with the DKG result sent by operator_2 would not report that operator_1 was inactive during result submission. It would be a transaction of Operator-Only type. It would be reimbursed (maybe during approval and not during submission, but still).
To report the inactivity of operator_1 and it's ineligibility for rewards a separate transaction would need to be sent by a willing party. That would be transaction of the Punishment type. This transaction would not be reimbursable.
Is that right so far?
If so, what would be the incentive here for posting of this second transaction? We're not slashing here, and funds saved by not paying the rewards to operator_1 will not increase the rewards for other operators, so why would anybody care to post that?

Author: Your understanding here is totally correct - I changed the incentive type to Operator-Only in [6878b4a](https://github.com/keep-network/tbtc-v2/pull/127/commits/6878b4add75c025371166f7982465278ebfc626d), and specified that we should be refunding eth.

So it ends up being an operator-only transaction that causes some punishment. Blurry lines, but the end result is that no one gets slashed tokens, only operators are able to submit it, and we need to reimburse the eth for whoever does, and we want to prevent the operators from frontrunning each other.

> <<punishment,Punishment>>.

Reviewer: I am not sure if such a transaction should be _punishment_ or _operator only_, I'd rather say the latter.

First, I do not think we should reward for this transaction. ETH reimbursement is fine but the submitter should be net-0. We do not want to incentivize this call, while we do want incentivize _punishment_ transactions. Second, only operators from the given group can assemble this transaction because it requires a sufficient number of signatures to support the motion. Such a motion needs to be proposed off-chain and signers need to work off-chain to share their signatures supporting it.

Author: addressed in [6878b4a](https://github.com/keep-network/tbtc-v2/pull/127/commits/6878b4add75c025371166f7982465278ebfc626d)

In order to compensate the operator for posting the transaction, the gas spent
will be reimbursed by a DAO-funded eth pool in the same transaction with
governable limits on how much eth can be spent on gas.

[[public-knowledge]]
==== Public-Knowledge
Public-Knowledge transactions are where anyone has access to the information
required to assemble the transaction.

In order to prevent wasting gas on racing to submit, we can either use an
off-chain gentleman's agreement for the operators like in
<<operator-only,Operator-Only>> transactions, or we can delegate the
transactions to a network like https://www.gelato.network/[Gelato]. We can't
cost-effectively stop members of the public from trying to race to submit.

To compensate these transactions, whoever posts them will have the gas spent
reimbursed by a DAO-funded eth pool in the same transaction, with governable

> limits on how much eth can be spent on gas.

Author: addressed in [6878b4a](https://github.com/keep-network/tbtc-v2/pull/127/commits/6878b4add75c025371166f7982465278ebfc626d) (accidentally stuck it in with another change - sorry!)

other wallet addresses is fraud, and is <<Punishment,punishable>>. Moving funds
without sending a follow-up proof is <<Punishment,punishable>>. Failure to
close a wallet that failed a <<heartbeat,heartbeat>> or fell below
`wallet_min_btc`, or exceeded `wallet_max_age` is <<Punishment,punishable>>.
Failure to complete the process before `wallet_closure_timeout` has elapsed is
<<Punishment,punishable>>.

Any unswept funds can be returned via the 30-day return script, though since
we're redeeming from the oldest wallet and only sweeping/depositing to the
newest wallets this hopefully won't come up often 🤞.

[[wallet-closure-transaction]]
===== Wallet Closure Transactions

Governable Parameters:

- `wallet_transfer_max`: The most amount of BTC a wallet can transfer to a
  single other wallet during a wallet closure.

When a wallet <<closing-a-wallet,closes>>, if there is any amount of bitcoin
remaining, it needs to be transfered to another _live_ wallet. Our input to the
transaction is the closing wallet's main UTXO, and we create an output for
every multiple of `wallet_transfer_max` we have in BTC in the closing wallet.

> The destination wallets are chosen for their public key hash's modulus distance

Author: If you don't do the modulus then numbers in the middle of the range are statistically more likely to be chosen, since they're closer to more numbers than numbers at the ends. Here's a simulated output:

```
const arr = [0,1,2,3,4,5,6,7]
let results = {}
for (let i = 0; i < 1000; i++) {
  const sample = getRandomSample(arr, 3)
  const option1DistanceSquared = Math.pow(sample[0] - sample[1], 2)
  const option2DistanceSquared = Math.pow(sample[0] - sample[2], 2)
  if (option1DistanceSquared < option2DistanceSquared) {
    results[sample[1]] = (results[sample[1]] || 0) + 1
  } else if (option1DistanceSquared > option2DistanceSquared) {
    results[sample[2]] = (results[sample[2]] || 0) + 1
  } else if (sample[1] > sample[2]){
    results[sample[1]] = (results[sample[1]] || 0) + 1
  } else {
    results[sample[2]] = (results[sample[2]] || 0) + 1
  }
}

console.log(results)
```
`getRandomSample` is grabbing a random sample from `[0,1,2,3,4,5,6,7]` like `[4,2,6]`. The first number is the wallet that's being closed and the second and third are the options. Here's how often each number is chosen without a modulus from 1000 runs:

```
{
  '0': 48,
  '1': 108,
  '2': 154,
  '3': 163,
  '4': 151,
  '5': 147,
  '6': 152,
  '7': 77
}
```

But with a modulus:

```
function dist(a, b, modulus) {
  let amod = a % modulus
  let bmod = b % modulus

  if (amod < bmod) {
    const temp = amod
    amod = bmod
    bmod = temp
  } 
  return Math.min(amod-bmod, bmod-amod+modulus);
}

const arr = [0,1,2,3,4,5,6,7]
let results = {}
for (let i = 0; i < 1000; i++) {
  const sample = getRandomSample(arr, 3)
  const option1Distance = dist(sample[0], sample[1], arr.length)
  const option2Distance = dist(sample[0], sample[2], arr.length)

  if (option1Distance < option2Distance) {
    results[sample[1]] = (results[sample[1]] || 0) + 1
  } else if (option1Distance > option2Distance) {
    results[sample[2]] = (results[sample[2]] || 0) + 1
  } else if (sample[1] > sample[2]){
    results[sample[1]] = (results[sample[1]] || 0) + 1
  } else {
    results[sample[2]] = (results[sample[2]] || 0) + 1
  }
}
console.log(results)
```

We get a nice random distribution
```
{
  '0': 122,
  '1': 111,
  '2': 120,
  '3': 105,
  '4': 119,
  '5': 128,
  '6': 159,
  '7': 136
}
```

to the closing wallet's public key hash (both represented as a number). Ties are
broken by the destination with the greater public key hash, and if we run out of
viable wallets, it is okay to reuse them (though we prefer not to).

For all the below examples, say that the public key hashes can range from 0 to
99 (mod 100). Wallet#43 Means that the PKH of the Wallet is 43. Real public key
hashes and their max mod values will be much higher in practice!

Exceeding Transfer Max Example:

- Wallet#90 with balance `250 BTC` needs to close.
- `wallet_transfer_max = 100 BTC`
- Live wallet options are Wallet#30, Wallet#80, Wallet#25, Wallet#15
- Calculate distances:

> - Wallet#30 = min(90 - 30, 130 - 90) = min(60, 40) = 40

Reviewer: We have two arguments passed to `min` because we want to look from both sides, right? That is, we can imagine the list of wallets as a circle and not as a line. Is it needed though? 

Thinking of it just as just a line would be a bit easier to implement. Are there reasons for doing it another way?

Author: Yup! Explained it in depth [above](https://github.com/keep-network/tbtc-v2/pull/127#discussion_r828731782)

- Wallet#42 = min(52 - 42, 142 - 52) = min(10, 90) = 10

Since Wallet#59 is the closest and Wallet#52's balance is below
`wallet_transfer_max`, we transfer the whole balance to Wallet#59.

Tiebreak Example:

- Wallet#52 with a balance `13 BTC` needs to close.
- `wallet_transfer_max = 100 BTC`
- Live wallet options are Wallet#13, Wallet#60, Wallet#44, Wallet#75
- Calculate distances:
- Wallet#13 = min(52 - 13, 113 - 52) = min(39, 61) = 39
- Wallet#60 = min(60 - 52, 152 - 60) = min(8, 92) = 8
- Wallet#44 = min(52 - 44, 144 - 52) = min(8, 92) = 8
- Wallet#75 = min(75 - 52, 152 - 75) = min(23, 77) = 23

Since Wallet#60 is tied with Wallet#44 for the closest, we favor Wallet#60
because 60>44. Since Wallet#52's balance is below `wallet_transfer_max`, we
transfer the whole balance to Wallet#60.

Wallet Reuse Example:

- Wallet#35 with a balance `413 BTC` needs to close.
- `wallet_transfer_max = 100 BTC`
- Live wallet options are Wallet#20, Wallet#30
- Calculate distances:
- Wallet#20 = min(35 - 20, 120 - 35) = min(15, 85) = 15
- Wallet#30 = min(35 - 30, 130 - 35) = min(5, 95) = 5

Deposit 100 BTC into Wallet#30, 100 BTC into Wallet#20, 100 BTC into Wallet#30,

> 100 BTC into Wallet#20, 13 BTC into Wallet#30 for a total of 213 BTC into Wallet#30 and 200 BTC into Wallet#20.

Reviewer: Do we expect the operators to commit to which wallets they are moving the funds before they actually do it?

Reviewer: Answer: no. This would not be possible. We will allow anyone to challenge their choice later.

closing it. If there are still funds remaining, the <<operator-only,first
operator>> posts a <<operator-only,reimbursable>> transaction to the ethereum
chain declaring an intention to close the wallet. This transaction includes a
timestamp hereby referred to as `move_funds_start_timestamp`.

That operator proposes a BTC fee (as in <<sweeping,sweeping>>) and the
operators use the wallet's public key hash to choose viable (it must also have
a passing heartbeat and be above `wallet_min_btc`, and not exceeded the
`wallet_max_age) wallet(s) to transfer the remaining funds to. They construct a
P2PKH transaction moving the wallet's main UTXO to that wallet. For more
details on exactly how a wallet is chosen and how the funds are split up see
<<wallet-closure-transaction,the dedicated section>>.

After this transaction is complete, <<public-knowledge,the public>> is able to
submit a SPV proof to let ethereum know that the funds were transferred
nonfraudulently. Later, a <<donate,donation>> can be made by governance to
handle any difference in outstanding v2 token supply and locked BTC due to
mining transfer fees.

> Later, <<public-knowledge,the public>> at any time can challenge the operator's
> wallet choice. They do so by showing that there existed a live, valid wallet
> which was created after `move_funds_start_timestamp` that _wasn't_ chosen _and_
> should have been. These challenges are incentivized to happen sooner rather

Reviewer: ```suggestion
Later, <<public-knowledge,the public>> at any time can challenge the operator's
wallet choice. They do so by showing that there exists a live, valid wallet
which was created before `move_funds_start_timestamp` that _wasn't_ chosen _and_
should have been. These challenges are incentivized to happen sooner rather
```

We can not prove on-chain there was a live wallet without additional costs of tracking timestamps of wallet status updates. But I think it is a typo given the created before/after part. What we can prove easily in the challenge: there is a live wallet, created before `move_funds_start_timestamp` and that wallet should be chosen but it wasn't.

Author: yeah - definitely meant before! [dd68fe7](https://github.com/keep-network/tbtc-v2/pull/127/commits/dd68fe7bb46201b906f6d41f3cbcca28fb9e907b)

- `wallet_closure_timeout`: The amount of time that a wallet has to
  successfully move its funds to another wallet and inform the ethereum chain
  before it is at risk of punishment.
- `wallet_max_age`: The oldest we allow a wallet to become before we transfer the funds
  to a randomly selected wallet.
- `wallet_min_btc`: The smallest amount of btc a wallet can hold before we
  attempt to close the wallet and transfer the funds to a randomly selected
  wallet.

When a wallet drops below the `heartbeat` threshold during a
<<heartbeat,heartbeat>> or ends a <<redemption,redemption>> with less than
`wallet_min_btc` remaining, or exceeds the `wallet_max_age`, the operators
begin the process of closing it. If there are still funds remaining, the
<<operator-only,first operator>> posts a <<operator-only,reimbursable>>
transaction to the ethereum chain declaring an intention to close the wallet.

That operator proposes a BTC fee (as in <<sweeping,sweeping>>) and the
operators use the hash of the previous intention-to-close transaction as a
source of randomness to choose viable (it must also have a passing heartbeat
and be above `wallet_min_btc`, and be within the `random_transfer_depth`)

> wallet to transfer the remaining funds to. They construct a P2PKH transaction

Reviewer 2: Should we send the funds to only one wallet from the set of `random_transfer_depth` the newest wallets or to all wallets from that set, in equal fractions? Piotr told me there were some conversations pointing to that second solution...
Sending to bigger set of wallets seems to be safer - if there would be some possibility for an operator who controls the closing wallet to abuse the mechanism of moving funds if they control the recipient wallet as well, then having more recipient wallets minimizes the amount of funds at risk.

Author: From a risk standpoint, I *greatly* prefer distributing the funds across the wallets. The main question is how much extra that costs in terms of sweeping transactions and ethereum proofs

moving the wallet's main UTXO to that wallet.

After this transaction is complete, <<public-knowledge,the public>> is able to
submit a SPV proof to let ethereum know that the funds were transferred
nonfraudulently. Later, a <<donate,donation>> can be made by governance to
handle any difference in outstanding v2 token supply and locked BTC due to
mining transfer fees.

*Notes*: Transferring the BTC to any address other than the P2PKH of one of the
other wallet addresses is fraud, and is <<Punishment,punishable>>. Moving funds
without sending a follow-up proof is <<Punishment,punishable>>. Failure to
close a wallet that failed a <<heartbeat,heartbeat>> or fell below
`wallet_min_btc`, or exceeded `wallet_max_age` is <<Punishment,punishable>>.
Failure to complete the process before `wallet_closure_timeout` has elapsed is

> <<Punishment,punishable>>.

Reviewer 2: To report that timeout, anyone would be able to post a transaction calling a `notifyWalletClosureTimeout` function (or something named similarly) and if that transaction would pass the validation, first notifier would receive some percentage of the slashed amount, right?
If so, wouldn't we need to have a governable param `wallet_closure_timeout_notifier_bonus_multiplier` for managing what part of slashed amount goes to notifier?
Or maybe we could have a common `notifier_bonus_multiplier` param for redemption request timeout and wallet closure timeout notifiers?

Author: Yup! Though in the interest of keeping this PR well-scoped, I think it's fine to just say that it's "punishable", and leave the details of that punishment to a follow-up PR

Reviewer 2: Ok, I'm fine with that :)

Any unswept funds can be returned via the 30-day return script, though since

> we're redeeming from the oldest wallet and only sweeping/depositing to the
> newest wallet this hopefully won't come up often 🤞.

Reviewer 2: ```suggestion
we're redeeming from the oldest wallet and only sweeping/depositing to the
newest wallets this hopefully won't come up often 🤞.
```

Author: [ffcbf40](https://github.com/keep-network/tbtc-v2/pull/127/commits/ffcbf40b22feb64fc64fada6704c1232dc89ef74)

Once a wallet no longer has funds and is not the primary wallet for new
deposits, it can be <<closing-a-wallet,closed>> and operators are no longer
required to maintain it.

[[closing-a-wallet]]
==== Closing A Wallet

Governable Parameters:

- `heartbeat`: The number of group members required for a
  <<heartbeat,heartbeat>> to be successful.
- `random_transfer_depth`: The depth (sorted youngest to oldest) at which we
  randomly select a new wallet to be the custodian of the old wallet's funds
  when closing a wallet. For instance, a depth of 3 means that the 3 youngest
  wallets can be randomly chosen.
- `wallet_closure_timeout`: The amount of time that a wallet has to
  successfully move its funds to another wallet and inform the ethereum chain
  before it is at risk of punishment.
- `wallet_max_age`: The oldest we allow a wallet to become before we transfer the funds
  to a randomly selected wallet.
- `wallet_min_btc`: The smallest amount of btc a wallet can hold before we
  attempt to close the wallet and transfer the funds to a randomly selected
  wallet.

When a wallet drops below the `heartbeat` threshold during a

> <<heartbeat,heartbeat>> or ends a <<redemption,redemption>> with less than

Author: [2c89353](https://github.com/keep-network/tbtc-v2/pull/127/commits/2c89353ae98c69f909cf510441b23f93df31f023)

`wallet_min_btc` remaining, or exceeds the `wallet_max_age`, the operators
begin the process of closing it. If there are still funds remaining, the
<<operator-only,first operator>> posts a <<operator-only,reimbursable>>
transaction to the ethereum chain declaring an intention to close the wallet.

That operator proposes a BTC fee (as in <<sweeping,sweeping>>) and the
operators use the hash of the previous intention-to-close transaction as a
source of randomness to choose viable (it must also have a passing heartbeat

> and be above `wallet_min_btc`, and be within the `random_transfer_depth`)

Author: [70e2369](https://github.com/keep-network/tbtc-v2/pull/127/commits/70e23696c47a148808b66445025474c9102792bf)

we're redeeming from the oldest wallet and only sweeping/depositing to the
newest wallets this hopefully won't come up often 🤞.

[[wallet-closure-transaction]]
===== Wallet Closure Transactions

Governable Parameters:

- `wallet_transfer_max`: The most amount of BTC a wallet can transfer to a

>   single other wallet during a wallet closure.

Author: That's what Promethea originally proposed, but I didn't see a compelling reason to entangle them (it confuses the naming and reduces governance flexibility) and scope-creeps the PR :D

  removed from reward eligibility after failing a heartbeat.
- `fraud_challenge_defend_timeout`: The amount of time the wallet has to defend
  against a fraud challenge.
- `fraud_challenge_deposit_amount`: The amount of ETH the party challenging the
  wallet for fraud needs to deposit.
- `fraud_notifier_reward_multiplier`: The percentage of the notifier reward
  from the staking contract the notifier of a fraud receives.
- `fraud_slashing_amount`: The amount of stake slashed from each member of a
  wallet for a fraud.
- `heartbeat_block_length`: The number of ethereum blocks until the next
  heartbeat. If set to 40, then the signers sign every 40th block.
- `heartbeat`: The number of group members required for a
  <<heartbeat,heartbeat>> to be successful.
- `max_gas_refund_price`: The highest amount of gwei that the gas refund
  contract will pay out per gas for a refund transaction.
- `random_transfer_depth`: The depth (sorted youngest to oldest) at which we
  randomly select a new wallet to be the custodian of the old wallet's funds
  when closing a wallet. For instance, a depth of 3 means that the 3 youngest
  wallets can be randomly chosen.

> - `redemption_request_timeout_redemeer_bonus_multiplier`: The percentage of the
>   notifier reward from the staking contract the redeemer receives in case of a
>   redemption timeout.
> - `redemption_request_timeout_redemeer_bonus_multiplier`: The percentage of the
>   notifier reward from the staking contract the redeemer receives in case of a
>   redemption timeout.
> - `redemption_request_timeout_slashing_amount`: The amount of stake slashed
>   from each member of a wallet for a timed-out redemption request.
> - `redemption_request_timeout_slashing_amount`: The amount of stake slashed
>   from each member of a wallet for a timed-out redemption request.
> - `redemption_request_timeout`: The amount of time the wallet has to provide
>   redemption proof.
> - `redemption_request_timeout`: The amount of time the wallet has to provide
>   redemption proof.
> - `redemption_treasury_fee`: The percentage of redeemed amount put aside as a
>   treasury fee.
> - `redemption_treasury_fee`: The percentage of redeemed amount put aside as a
>   treasury fee.

Reviewer 2: Some parameters got duplicated
```suggestion
- `redemption_request_timeout_redemeer_bonus_multiplier`: The percentage of the
  notifier reward from the staking contract the redeemer receives in case of a
  redemption timeout.
- `redemption_request_timeout_slashing_amount`: The amount of stake slashed
  from each member of a wallet for a timed-out redemption request.
- `redemption_request_timeout`: The amount of time the wallet has to provide
  redemption proof.
- `redemption_treasury_fee`: The percentage of redeemed amount put aside as a
  treasury fee.
```
